File: Golang/530. Minimum Absolute Difference in BST.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func getMinimumDifference(root *TreeNode) int {
    values := make([]int, 0, 100)
    search(root, &values)
    ans := distance(values[0], values[1])
    i := 0
    for i < len(values)-1 {
        ans = min(ans, distance(values[i], values[i+1]))
        i++
    }
    return ans
}

func search(node *TreeNode, values *[]int) {
    if node != nil {
        search(node.Left, values)
        *values = append(*values, node.Val)
        search(node.Right, values)
    }
}

func min(a, b int) int {
    if a > b {
        return b
    } else {
        return a
    }
}

func distance(a, b int) int {
    if a > b {
        return a - b
    } else {
        return b - a
    }
}

File: Python3/530. Minimum Absolute Difference in BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        values = [] # 120ms 将二叉树转化为了有序数组
        def Tree2List(node, values):
            if node:
                Tree2List(node.left, values)
                values.append(node.val)
                Tree2List(node.right, values)
        Tree2List(root, values)
        return min([abs(x-y) for x, y in zip(values[:-1], values[1:])])            
